main.py

```diff
@@ -670,6 +670,42 @@ def delete_link_metadata(self, request: Request) -> JSONResponse:
         self.notify_topology_update()
         return JSONResponse("Operation successful")
 
+    @rest('v3/links/{link_id}', methods=['DELETE'])
+    def delete_link(self, request: Request) -> JSONResponse:
+        """Delete a disabled link from topology.
+         It won't work for link with other statuses.
+        """
+        link_id = request.path_params["link_id"]
+        try:
+            with self._links_lock:
+                link = self.links[link_id]
+                if link.status != EntityStatus.DISABLED:
+                    raise HTTPException(409, detail="Link is not disabled.")
+                link = self.links.pop(link_id)
+                switch_a = link.endpoint_a.switch
+                switch_b = link.endpoint_b.switch
+                if link.endpoint_a.link and link == link.endpoint_a.link:
+                    link.endpoint_a.link = None
+                    link.endpoint_a.nni = False
+                if link.endpoint_a.link and link == link.endpoint_b.link:
+                    link.endpoint_b.link = None
+                    link.endpoint_b.nni = False
+        except KeyError:
+            raise HTTPException(404, detail="Link not found.")
+
+        self.topo_controller.upsert_switch(switch_a.id, switch_a.as_dict())
+        if switch_a != switch_b:
+            self.topo_controller.upsert_switch(
+                switch_b.id, switch_b.as_dict()
+            )
+        self.topo_controller.delete_link(link_id)
```

Comment:
@Alopalao, overall looks good, but I've just also realized a potential corner case here where DB connectivity is temporarily lost, and let's say the auto retries still don't succeed either (this is a very rare case considering the current prod deployment at AmLight at least, but it can happen), so in that case `self.links` would got popped, but some of the DB ops (write or deletion) might not get completed, that said, the safest options here is indeed acquire `self._links_lock` as you done, but still hold it while performing the DB ops and only pop the link after this delete operation here, see what I mean? Misbehaving clients might still slow things down since the lock will keep getting acquired, but that's not too bad, considering the other DB inconsistent problems that we might end up in a very worst case scenario (which would require us to deal with a new release patch to fix it or DB script, so let's try to avoid this in the first place). Regarding misbehaving clients potentially making too much request we can rate limit them in the future, so it's OK that we also have a solution for it in the future.

Reply:
Understood, moved the DB updates into the lock.